core/src/main/java/org/apache/iceberg/BaseEntriesTable.java

```diff
@@ -125,31 +130,128 @@ ManifestFile manifest() {
 
     @Override
     public CloseableIterable<StructLike> rows() {
+      Types.NestedField readableMetricsField = projection.findField(MetricsUtil.READABLE_METRICS);
       // Project data-file fields
       CloseableIterable<StructLike> prunedRows;
-      if (manifest.content() == ManifestContent.DATA) {
-        prunedRows =
-            CloseableIterable.transform(
-                ManifestFiles.read(manifest, io).project(fileSchema).entries(),
-                file -> (GenericManifestEntry<DataFile>) file);
+      if (readableMetricsField == null) {
+        if (manifest.content() == ManifestContent.DATA) {
```

Comment:
Can we call helper method like entries() as you wrote below?  We can have entries() return CloseableIterable<GenericManifestEntry<? extends ContentFile<?>>>, and then do another CloseableIterable cast here from GenericManifestEntry to StructLike, which should be safe.

Reply:
yeah make sense! Also realized from other PR that this can be simplified further with 

`ManifestFiles.open(manifest, io, specsById)` so it's agnostic read to data and delete manifests